src/celeritas/neutron/model/NeutronInelasticModel.cc

```diff
@@ -0,0 +1,237 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/neutron/model/NeutronInelasticModel.cc
+//---------------------------------------------------------------------------//
+#include "NeutronInelasticModel.hh"
+
+#include "corecel/math/ArrayOperators.hh"
+#include "corecel/math/Quantity.hh"
+#include "celeritas/global/CoreParams.hh"
+#include "celeritas/global/CoreState.hh"
+#include "celeritas/io/ImportPhysicsVector.hh"
+#include "celeritas/mat/MaterialParams.hh"
+#include "celeritas/phys/InteractionApplier.hh"
+#include "celeritas/phys/PDGNumber.hh"
+#include "celeritas/phys/ParticleParams.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from model ID and other necessary data.
+ */
+NeutronInelasticModel::NeutronInelasticModel(ActionId id,
+                                             ParticleParams const& particles,
+                                             MaterialParams const& materials,
+                                             ReadData load_data)
+{
+    CELER_EXPECT(id);
+    CELER_EXPECT(load_data);
+
+    HostVal<NeutronInelasticData> data;
+
+    // Save IDs
+    data.scalars.action_id = id;
+    data.scalars.neutron_id = particles.find(pdg::neutron());
+
+    CELER_VALIDATE(data.scalars.neutron_id,
+                   << "missing neutron particles (required for "
+                   << this->description() << ")");
+
+    // Save particle properties
+    data.scalars.neutron_mass = particles.get(data.scalars.neutron_id).mass();
+
+    // Load neutron inelastic cross section data
+    make_builder(&data.micro_xs).reserve(materials.num_elements());
+    for (auto el_id : range(ElementId{materials.num_elements()}))
+    {
+        AtomicNumber z = materials.get(el_id).atomic_number();
+        this->append_micro_xs(load_data(z), &data);
+    }
```

Comment:
We could use the `GenericGridBuilder` to simplify here:
```suggestion
    CollectionBuilder micro_xs{&data.micro_xs};
    GenericGridBuilder build_grid{&data.reals};
    for (auto el_id : range(ElementId{materials.num_elements()}))
    {   
        AtomicNumber z = materials.get(el_id).atomic_number();
        micro_xs.push_back(build_grid(load_data(z)));
    }
```

Reply:
Nice! Did not know this feature.  Changed.